Ladder step3

clown0425 opened the pull request:
플레이어 이름을 입력받는 구조로 수정
접근 제어 적용

clown0425 committed:
userinput파일을 삭제하고 Inputview파일 생성
플레이어 이름과 사다리 높이를 입력받는 역할

clown0425 committed:
에러 메세지

clown0425 committed:
requestMassage

clown0425 committed:
playername print를 위한 nameBoard 생성을 위한 구조체

clown0425 committed:
출력할 사다리를 데이터화

clown0425 committed:
이름을 적절히 출력하기 위해
ladders크기를 한칸 늘림

clown0425 committed:
플레이어 이름과 사다리 출력

clown0425 committed:
ladderPlayer

clown0425 committed:
ladderPlayer구조체 활용

godrm commented on LadderGame/LadderGame/nameBoard.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+
+struct nameBoard {
+    var nameBoard  = Array<String>()
```
`struct nameBoard` 처럼 타입 이름은 대문자로 시작하세요.
내부 변수는 private 으로 지정해서 외부에서 값을 바꾸지 못하도록 만드세요.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -1,44 +1,37 @@
-    lazy var Ladders:[[Character]] = Array(repeating:Array(repeating: " ", count:ladderWidthLength), count:ladderHeight)
+    private lazy var widthLadderIsExistence:[[Bool]] = Array(repeating:Array(repeating: false, count:playerNumber-1), count:ladderHeight)
+    
+    lazy var Ladders:[[String]] = Array(repeating:Array(repeating: Ladder.emptyStep.rawValue, count:ladderWidthLength+1), count:ladderHeight)
```
`lazy var`를 사용했네요. 어떤 용도로 선택한 건가요?

clown0425 replied:
사실 저도 이부분에 대해 의문이 들어 코드를 쓸 당시 여러 가지로 찾아 보았는데..
확실한 결론에 다다르지는 못했습니다
초기화 중에는 다른 인스턴스 메서드 호출이 불가능 하다
메서드 호출중 초기화되지 않은 프로퍼티를 사용할 수도 있기 때문에 초기화를 먼저 완료하지 않으면 컴파일 에러가 발생한다
위의 말을 기준으로 어렴풋이 이해한 바는 아래와 같습니다
플레이어 수와 사다리 높이에 따라 다른 값이 계산되어 저장 되는데
초기화 되지 않은 값을 이용하기 때문에 초기화가 완료된후 호출할때 값이 계산되어 저장되는 형식을  쓴다

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -1,44 +1,37 @@
-    var widthLadderNumber:Int {
+    private var widthLadderNumber:Int {
         return Int.random (in: 1...(playerNumber - 1) / 2 * ladderHeight)
     }
```
widthLadderNumber 는 왜 computed variable 로 만든건가요? 
매번 새로 계산해야 하는 이유가 있을까요?

clown0425 replied:
이전 버전에서 사용자에게 여러번 입력 받을 수도 있다는 가정하에 쓰던걸 그대로 썻네요..
수정했습니다

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -83,31 +76,31 @@ struct ladderBoard {
+    private mutating func loopOfPlayerNumberForHeight(_ i : Int) {
         for j in 0..<playerNumber{
-            Ladders[i][j*2] = "ㅣ"
+            Ladders[i][j*2+1] = Ladder.bar.rawValue
```
사다리 데이터 정보를 만들 때 꼭 `|` bar를 명시해야 할까요? 
이 정보는 사다리를 표현하는데 최소 데이터가 아니라 출력을 위한 정보일 뿐입니다.

clown0425 replied:
무슨 말인지 잘 모르겠습니다 
고민해 보겠습니다

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -120,3 +113,4 @@ struct ladderBoard {
 }
```
보통 init() 처럼 객체를 생성하는 메소드는 맨위쪽에 선언해서 코드를 읽을 때 생성하는 시점부터 시간흐름 순서대로 읽을 수 있도록 배려하는 편입니다. 참고하세요.

godrm commented on LadderGame/LadderGame/Inputview.swift:
```diff
@@ -0,0 +1,110 @@
+        if let convertedString = optionalString{
+            return convertedString
+        }
+        return "fail"
```
에러 / 실패한 값을 리턴 타입에 특정한 값으로 넘기지 마세요. 
대신 명확하게 Error를 throw 하거나 다른 방법으로 처리하세요.

godrm commented on LadderGame/LadderGame/Inputview.swift:
```diff
@@ -0,0 +1,110 @@
+    private func printErrorMassage(_ isCorrect:Bool, _ errorMassage:ErrorMassage) {
+        if isCorrect == false {
+            print(errorMassage.rawValue)
+        }
```
InputView 에서 실패에 대한 문구를 직접 처리하지 마세요. 
Input을 호출하는 상위 모듈에서 결정하세요.

godrm commented on LadderGame/LadderGame/Inputview.swift:
```diff
@@ -0,0 +1,110 @@
+        var isCorrectNumber:Bool
+        repeat{
+            isCorrectNumber = checkNumberOfHeight()
+        }while (isCorrectNumber) == false
```
InputView 내부에서 반복문으로 흐름 제어를 하지마세요.
대신 InputView 상위 모듈에 정확한 에러코드를 전달하고,
반복해서 호출할지 끝낼지는 상위 모듈에서 판단하세요.

godrm commented on LadderGame/LadderGame/Inputview.swift:
```diff
@@ -0,0 +1,110 @@
+    
+    var playerNumber:Int{
+        return playerNames.count
+    }
```
playerNames 나 ladderHeight 를 직접 가져가기 보다는
아래 readNames() 나 readLadderHeight() 에서 리턴하면 되지 않을까요?
변수에 직접 접근하지 않는게 더 좋은 방식입니다.

godrm reviewed (commented):
구조체를 만들때 관련있는 코드들을 응집도 있게 합쳐보세요.

clown0425 committed:
구조체 이름 수정
-소문자로 시작하던 부분을 대문자로 변경

clown0425 committed:
nameBoard변수에 접근을 제한
drawPlayerNameBoard함수에서 nameBoard를 반환하여 활용하도록 수정

clown0425 committed:
ladders배열에 대한 접근 제한
사다리를 그리는 두 함수 drawAllWidthLadder drawAllHeightLadder 는 같은 추상화 수준으로
하나의 함수drawLadder로 묶음
drawLadder에서 ladders를 반환하는 흐름으로 수정

clown0425 committed:
readNames,readLadderHeight 에서 playerNames,ladderHeight를 반환하여 진행하는 방식으로 수정
따라서playerNames,ladderHeight에 대한 접근 제한
playerNumber 연산 프로퍼티가 불필요 하여 삭제

clown0425 committed:
오류를 던저 메인에서 처리할수 있게 수정

clown0425 committed:
init함수 위치 수정

clown0425 commented:
-변수에 접근하는 방식이 아닌 함수가 리턴하는 방식으로 수정
-상위모듈에 에러를 전달하여 처리하는 방식으로 수정

godrm commented on LadderGame/LadderGame/Inputview.swift:
```diff
@@ -1,77 +1,61 @@
+        if let convertedString = optionalString{
+            return convertedString
+        }
+        throw InputError.isNotString
```
이정도가 꼭 메소드로 있어야 할 정도일까요?
`??` 나 guard else 구문 수준으로 해결이 안되는 정도일까요?

clown0425 replied:
1.자주 사용해야 한다면 메소드로 만들어 놓고 쓰는것이
  매번 ?? 나 guard else를 사용하는 것보다 편리하고 
 코드를 읽을 때도 직관적이어 보인다고 생각합니다

2.오류를 던져야 할때 좀더 쉽게 던질수 있다고 생각합니다

메소드로 만들어 쓰게 되면 생기는 단점이나
??  //  guards else 구문을 쓰게 되면 생기는 장점이 있을까요?

godrm replied:
같은 기능을 함수(메소드)로 호출하는 것과 그대로 표현할 수 있는게 있다면 얼마나 자주 반복되는 지 판단해 볼 것 같습니다. 
이 함수 내부에 작성한 코드 조각이 자주 반복된다면 분리해도 좋을 것 같습니다. 
함수 호출이 코드 조각보다는 성능이 느려지지만 그런 부분을 고민하면 `@inlining` 방식도 있습니다. 
동일한 guard else 가 여기저기 반복되는 경우가 아니라면 꼭 있어야 하는지 확인한 겁니다.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -1,6 +1,11 @@
+    
     private let playerNumber:Int
     private let ladderHeight:Int
     
```
init() 함수 위치를 맨위로 바꾼 것은 좋습니다.
프로퍼티 속성과는 어떤 게 더 위에 있는게 좋을까요? 
init()에서는 무엇을 채우게 되죠? 본인의 기준은 무엇인가요?

clown0425 replied:
이 질문 때문에 객체에 대해 많이 생각해 보게 되었습니다
초기화 하는 값은 각각의 인스턴스에 특징을 부여 하는 거라는 생각이 들었습니다
입력된 값에 따라 다른 기능을 하고 다른 결과를 내놓기 때문입니다
그렇게 생각이 꼬리를 물고
객체가 존재해야 하는 이유 ,역할분담 이란 무엇일까 등에 대해 고민해 보았습니다 
아직 어떤 답을 찾지는 못했습니다 
더 고민해서 정리해 보겠습니다

godrm replied:
좋은 꼬리 질문이네요. 너무 모든 것을 한꺼번에 정리하려고 하기 보다는 작은 문제로 나눠서 답을 찾아가보세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -9,20 +9,30 @@
+    }catch let error as InputError{
+        print(error.description)
+    }catch{
+        print("알 수 없는 에러")
```
여기 에러 문구도 InputError 코드로 표현할 수 있지 않을까요?
값을 하드코딩하기 보다는 코드로 표현하는게 좋습니다.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -78,19 +83,19 @@ struct ladderBoard {
     private mutating func loopOfPlayerNumberForHeight(_ i : Int) {
         for j in 0..<playerNumber{
-            Ladders[i][j*2+1] = Ladder.bar.rawValue
+            ladders[i][j*2+1] = Ladder.bar.rawValue
```
`ladders` 사다리를 표현하는 데이터 구조에 대해 두 가지 측면이 있습니다.
- 사다리 발판 하나를 표현하는 데 `String` 타입이 최적의 타입일까요? 발판이 있다 없다는 정보만 필요하다면 Bool 타입이 적합하지 않을까요?
- 사다리 발판 데이터 구조는 출력을 위한게 아니라 사다리를 최소한의 메모리 공간을 써서 효과적으로 표현하는 게 좋습니다. 그런 맥락에서 `Ladder.bar`를 데이터 구조에 포함할 필요는 없습니다. 출력할 때만 붙여서 표시해도 되는거니까요.

clown0425 committed:
값을 하드코딩 하지 않고 IputError코드로 대체

clown0425 committed:
1.가로 사다리 존재 여부를 결정하는 함수 수정
2.사다리를 그리는 구조체와 출력하는 구조체 명확히 구분
 -세로 사다리를 데이터 구조에 저장하지 않고 출력만 하게 수정
 -플레이어 이름을 출력하는 구조체도 위와 같은 이유로 삭제후 결과만 출력

clown0425 commented:
사다리를 표현하는 데이터 구조에 대해 생각해 보고 코드를 수정했습니다
-출력에만 필요한 요소는 저장하지 않고 출력만 하게 수정했습니다
-랜덤으로 가로 사다리를 생성하는 함수를 수정했습니다

godrm commented:
이렇게 해서 데이터 구조도 단순해지고, 출력 과정도 단순해진 것 같습니다.
다음 스탭에서는 지금 구성한 구조체와 함수들을 아주 작은 단위로 테스트를 하는 겁니다.
어떤 기능을 성공하는지, 실패하는지, 어디까지 테스트해야 하는지 학습하고 진행해보세요.
